github: calculate sleep time based on Github server time

aliculPix4D opened the pull request:
- use date header reporting the server time to calculate the sleep time
- add a jitter time to calculated sleep time to prevent creating a Thundering Herd
- prevent negative sleep times. Minimal sleep time should be equal to jitter time

aliculPix4D commented on github/commitstatus.go:
```diff
@@ -244,15 +254,23 @@ func checkForRetry(res httpResponse, waitTime, maxSleepTime time.Duration) time.
 		} else {
-			return sleepTime
+			// Don't allow negative sleep times. Minimum sleep time should be equal to jitter time.
+			return max(sleepTime, jitter)
```
if we need to sleep, the minimal sleep time will a random number between 0-30.

marco-m-pix4d replied:
I now start to think that this line is wrong. Above we just added jitter to sleepTime, here there should be no need to calculate any max, we should just return sleepTime...

aliculPix4D replied:
Yes, I agree with you but I wanted to ensure that no negative time can happen anymore.

So my thinking was that this line:
```
sleepTime := res.rateLimitReset.Sub(res.date) 
```
can still return anything if there is any change on the server side. So adding 30sec to something negative and larger than 30sec can still create a negative sleeptime.

It should never happend of course, and `res.rateLimitReset.Sub(res.date) ` should really always be a positive value, but I was sceptical... 

So I wanted to ensure we really can't have the negative time anymore...hence `max(sleeptime, jitter)`

I am fine with removing it, just wanted to make it clear that:
- I did this on purpose
- mostly in "panic", "paranoid" mode trying to ensure that we don't have negative sleep times anymore

marco-m-pix4d replied:
I see. The problem with that approach is that the intent was hidden, so a reader is left thinking that the code is wrong. Now that I know the intent, I can advice.

From my understanding, the scenario you are mentioning can happen only in case of a clock severely misbehaving either on the server or client side. If we start going down this slope, we will never finish with "what if" scenarios, adding complexity to the code.

For these reasons, this max should be removed, eventually replaced by a log.Warning that an unexpected negative time has been detected. This log line should be added at the first moment where we can detect the problem, when we calculate sleepTime, before adding any jitter.

aliculPix4D replied:
Still to do: 
- [ ] add warn log line if sleep time is negative

aliculPix4D commented on github/commitstatus.go:
```diff
@@ -178,6 +182,12 @@ func httpRequestDo(req *http.Request) (httpResponse, error) {
+
+	// time.RFC1123 is the format of the Github date header
+	// example: Date:[ Mon, 02 Jan 2006 15:04:05 MST ]
+	date, _ := time.Parse(time.RFC1123, resp.Header.Get("date"))
```
So the date header is always reported and its in `Mon, 02 Jan 2006 15:04:05 MST` format. I can't find anything about this header in Github documentation so I am worried with this a bit. Hence, on L112 I added `req.Header.Set("X-GitHub-Api-Version", "2022-11-28")` to at least try to ensure this doesn't start to break if there is some changes from Github side.

odormond replied:
I'm not surprised you don't see it in the GitHub doc because this is a standard HTTP header. See https://datatracker.ietf.org/doc/html/rfc2616#section-14.18.

marco-m-pix4d replied:
To add to Olivier's comment, if an header doesn't begin with "X-" AND the web server is well-behaved, for sure the header is a standard header and so is defined in an RFC.

If an header begins with "X-", then it depends. It can be something invented by the specific service (so in this case GtiHub) OR it could be still a standard header.

A very good starting point is https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers

aliculPix4D replied:
I added the link as comment.

marco-m-pix4d commented on CHANGELOG.md:
```diff
@@ -5,8 +5,20 @@ All notable changes to this project will be documented in this file.
+
+### Added
+
+- Adds a jitter time to calculated sleep time to prevent creating a Thundering Herd.
```
- Prevent being rate limited by GitHub: add ...

marco-m-pix4d commented on github/commitstatus.go:
```diff
@@ -107,6 +109,7 @@ func (s CommitStatus) Add(sha, state, targetURL, description string) error {
 	req.Header.Set("Authorization", "token "+s.token)
 	req.Header.Set("Accept", "application/vnd.github.v3+json")
 	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("X-GitHub-Api-Version", "2022-11-28")
```
per the review comments, this is not needed.

aliculPix4D replied:
Removed as suggested.

marco-m-pix4d commented on github/commitstatus.go:
```diff
@@ -178,6 +182,12 @@ func httpRequestDo(req *http.Request) (httpResponse, error) {
 	rateLimitReset, _ := strconv.Atoi(resp.Header.Get("x-ratelimit-reset"))
 	response.rateLimitReset = time.Unix(int64(rateLimitReset), 0)
+
+	// time.RFC1123 is the format of the Github date header
```
time.RFC1123 is the format of the HTTP `Date` header

aliculPix4D replied:
Changed as suggested.

marco-m-pix4d commented on github/commitstatus.go:
```diff
@@ -244,15 +254,23 @@ func checkForRetry(res httpResponse, waitTime, maxSleepTime time.Duration) time.
 			return 0
 		} else {
-			return sleepTime
+			// Don't allow negative sleep times. Minimum sleep time should be equal to jitter time.
```
The comment is somewhat misleading:
We avoid negative sleep time only by using the server's notion of clock, the jitter is unrelated, but the comment is about the max, which is unrelated to negative values, because sleepTime is guaranteed to be non negative.

So I would remove the comment entirely.

marco-m-pix4d replied:
See the comment above on how to make it clearer what is happening

aliculPix4D replied:
Fixed as suggested in the comment below.

marco-m-pix4d commented on github/commitstatus.go:
```diff
@@ -244,15 +254,23 @@ func checkForRetry(res httpResponse, waitTime, maxSleepTime time.Duration) time.
 	// If you exceed the rate limit, the response will have a 403 status and the x-ratelimit-remaining header will be 0
 	// https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit
 	case res.statusCode == http.StatusForbidden && res.rateLimitRemaining == 0:
-		sleepTime := time.Until(res.rateLimitReset)
+		sleepTime := res.rateLimitReset.Sub(res.date) + jitter
```
Here we are conflating two different operations in one line and we are missing a comment explaining what is happening. We could do instead (untested):

```go
// Calculate sleeptime based solely on the server clock. This is unaffected by the clock drift between server and client.
sleepTime := res.rateLimitReset.Sub(res.date)
// Be a good netizen by adding some jitter to the time we sleep.
sleepTime := res.rateLimitReset.Sub(res.date) + jitter
```

aliculPix4D replied:
Changed as suggested.

marco-m-pix4d commented:
@aliculPix4D Let me know when you have gone through all the comments; if anything unclear, we can also have a chat if you prefer.

aliculPix4D commented on CHANGELOG.md:
```diff
@@ -5,8 +5,20 @@ All notable changes to this project will be documented in this file.
+
+### Fixed
+
+- Fixes a bug with negative times that could occur due to the small difference between local (client) time and Github server time. Sleep time is now calculated using the date header (time as seen by the Github server). Additionally, negative sleep times are prevented by setting a minimum sleep time equal to the jitter time.
```
Changed as suggested.

aliculPix4D commented on cogito/ghcommitsink.go:
```diff
@@ -38,6 +39,8 @@ func (sink GitHubCommitStatusSink) Send() error {
 		MaxRetries:   maxRetries,
 		WaitTime:     waitTime,
 		MaxSleepTime: maxSleepTime,
+		// adds some randomness to sleep time to prevent creating a Thundering Herd
```
Removed as suggested.

aliculPix4D commented:
@marco-m-pix4d I addressed all your comments. Can you have a look again? 
I am espacially curious to hear your opinion on: https://github.com/Pix4D/cogito/pull/122/commits/26d0296e03f5c1f5c5af75393a7d662bc55c3dff

this is needed only to address your comment about negative sleep time:
>  log.Warning that an unexpected negative time has been detected

aliculPix4D commented on github/commitstatus.go:
```diff
@@ -111,18 +113,23 @@ func (s CommitStatus) Add(sha, state, targetURL, description string) error {
+		timeToSleep = checkForRetry(response, s.target.WaitTime, s.target.MaxSleepTime, s.target.Jitter)
+		switch {
+		case timeToSleep == 0:
+			break retryLoop
```
@marco-m-pix4d I didn't find a better way to break a loop. Is this in go language spirit?

marco-m-pix4d replied:
if you use an `if` as opposed to a `switch`, you can use the `break` as before without the label at line 116, although the code as-is is correct.

But, I have the feeling we can simplify all this. Let me think a moment.

aliculPix4D commented on github/commitstatus.go:
```diff
@@ -111,18 +113,23 @@ func (s CommitStatus) Add(sha, state, targetURL, description string) error {
+		case timeToSleep == 0:
+			break retryLoop
+		case timeToSleep < 0:
+			s.log.Warn("Sleep time should not have a negative value", "time", timeToSleep)
```
I was also thinking if we should just return an error here

marco-m-pix4d replied:
```go
s.log.Warn("unexpected: negative sleep time", "time", timeToSleep)
```

aliculPix4D commented on github/commitstatus.go:
```diff
@@ -175,9 +183,16 @@ func httpRequestDo(req *http.Request) (httpResponse, error) {
+	// time.RFC1123 is the format of the HTTP Date header
+	// example: Date:[ Mon, 02 Jan 2006 15:04:05 MST ]
+	// https://datatracker.ietf.org/doc/html/rfc2616#section-14.18
+	date, _ := time.Parse(time.RFC1123, resp.Header.Get("Date"))
```
same comment should apply here also: https://github.com/Pix4D/fly_helper/pull/158/files#r1201798389

> [marco-m-pix4d](https://github.com/marco-m-pix4d)(Marco Molteni) [2 minutes ago](https://github.com/Pix4D/fly_helper/pull/158/files#r1201798389)
In Go, never skip handling errors, unless there is a very specific reason.
This is equivalent to intercepting all exceptions in Python and dropping them.

aliculPix4D commented on github/commitstatus.go:
```diff
@@ -175,9 +183,16 @@ func httpRequestDo(req *http.Request) (httpResponse, error) {
-	response.rateLimitRemaining, _ = strconv.Atoi(resp.Header.Get("x-ratelimit-remaining"))
-	rateLimitReset, _ := strconv.Atoi(resp.Header.Get("x-ratelimit-reset"))
+	response.rateLimitRemaining, _ = strconv.Atoi(resp.Header.Get("X-RateLimit-Remaining"))
+	rateLimitReset, _ := strconv.Atoi(resp.Header.Get("X-RateLimit-Reset"))
```
same comment should apply here also: https://github.com/Pix4D/fly_helper/pull/158/files#r1201798389

> [marco-m-pix4d](https://github.com/marco-m-pix4d)(Marco Molteni) [2 minutes ago](https://github.com/Pix4D/fly_helper/pull/158/files#r1201798389)
In Go, never skip handling errors, unless there is a very specific reason.
This is equivalent to intercepting all exceptions in Python and dropping them.

aliculPix4D replied:
This is not an excuse in any case, but I "learned" the best practice from:
https://github.com/google/go-github/blob/f96b64ece988554ba6444cb305e6db1132a61454/github/github.go#L674

marco-m-pix4d replied:
From the documentation:

> The errors that ParseInt returns have concrete type *NumError and include err.Num = s. If s is empty or contains invalid digits, err.Err = ErrSyntax and the returned value is 0; if the value corresponding to s cannot be represented by a signed integer of the given size, err.Err = ErrRange and the returned value is the maximum magnitude integer of the appropriate bitSize and sign. 

So in case of error, the outcome of strconv.Atoi is `0`. They might have decided that `0` is OK, or they might have made a wrong assumption.

aliculPix4D replied:
All errors are now properly handled.

aliculPix4D commented:
@Pix4D/integration request for a re-review. Thanks

aliculPix4D commented on github/commitstatus.go:
```diff
@@ -175,9 +181,29 @@ func httpRequestDo(req *http.Request) (httpResponse, error) {
-	response.rateLimitRemaining, _ = strconv.Atoi(resp.Header.Get("x-ratelimit-remaining"))
-	rateLimitReset, _ := strconv.Atoi(resp.Header.Get("x-ratelimit-reset"))
-	response.rateLimitReset = time.Unix(int64(rateLimitReset), 0)
+	if limit := resp.Header.Get("X-RateLimit-Remaining"); limit != "" {
```
I took some inspiration from: https://github.com/google/go-github/blob/9e4ce880cd4db6c8f8459806a8c1276a14db3983/github/github.go#L665

It seems that its perfectly valid and possible that these two headers (`X-RateLimit-Remaining`, `X-RateLimit-Reset`) are not reported. But if they are reported and we fail to convert their value to integer we must report an error.
In the `go-github` pkg they are happy enough with return value of 0 and ignore the error, I prefer to be explicit about it.

marco-m-pix4d commented on github/commitstatus.go:
```diff
@@ -175,9 +181,29 @@ func httpRequestDo(req *http.Request) (httpResponse, error) {
-	rateLimitReset, _ := strconv.Atoi(resp.Header.Get("x-ratelimit-reset"))
-	response.rateLimitReset = time.Unix(int64(rateLimitReset), 0)
+	if limit := resp.Header.Get("X-RateLimit-Remaining"); limit != "" {
+		v, err := strconv.Atoi(limit)
```
`v` is really too short also if this is Go, `val` is OK.
I find the "failed to convert" confusing: as a user (so not seeing the code), what does it mean?
It is enough to say: 
```go
fmt.Errorf("X-RateLimit-Remaining: %s", err)
```
which would print:
```
X-RateLimit-Remaining: strconv.Atoi: parsing "foo": invalid syntax
```
As proof, see https://go.dev/play/p/6K64il-jIE8

marco-m-pix4d replied:
Consider also that strconv.Atoi is documented to return 0 in case of error, so we could simply write
```go
if limit := resp.Header.Get("X-RateLimit-Remaining"); limit != "" {
		response.rateLimitRemaining, err = strconv.Atoi(limit)
		if err != nil {
			return httpResponse{}, fmt.Errorf("X-RateLimit-Remaining: %s", err)
		}
}
```
or do as 1 line:
```
    // strconv.Atoi returns 0 in case of error.
	response.rateLimitRemaining, _ = strconv.Atoi(resp.Header.Get("X-RateLimit-Remaining"))
```

in the 3 cases, I think the result will be the same, AS LONG as this function is called on `response` only once. If it is called multiple times, then we must pick the correct behavior.

I don't have a strong opinion, as long as the code is consistent.

aliculPix4D replied:
> Consider also that strconv.Atoi is documented to return 0 in case of error

but not always :)

```
val, err := strconv.Atoi("9223372036854775808")
err = fmt.Errorf("X-RateLimit-Remaining: %s", err)
fmt.Println(val, err)
```
results in:
`9223372036854775807 X-RateLimit-Remaining: strconv.Atoi: parsing "9223372036854775808": value out of range`

marco-m-pix4d replied:
The example I gave was in case the header key was missing, which would result in a 0.

This code:
```go
if limit := resp.Header.Get("X-RateLimit-Remaining"); limit != "" {
    response.rateLimitRemaining, err = strconv.Atoi(limit)
    if err != nil {
        return httpResponse{}, fmt.Errorf("X-RateLimit-Remaining: %s", err)
    }
}
```
Can always be replaced by this:
```go
response.rateLimitRemaining, err = strconv.Atoi(limit)
if err != nil {
    return httpResponse{}, fmt.Errorf("X-RateLimit-Remaining: %s", err)
}
```

Regarding the out of range behavior, clamping to the biggest value encodable in the underlying int, seems quite reasonable to me, and is documented as such.

In any case, we need to keep the focus on the intent and not get lost on hypothetical corner cases.

What is the intent and the effect of returning in case of error?
Returning `httpResponse{}` will give the zero value to field `rateLimitRemaining`, which for ints is 0.
Is this what is wanted in case of error? Or is this expected not to be used because the caller is expected to handle the error and not look at `httpResponse` ?
